### DISCUSSION TITLE
EPOCH in moving frame

### DISCUSSION THREAD

**Discussion by DanRRRR** [epochpic/epoch#542] [General]
Some simulations need to be performed in elongated volume where for example short laser pulse propagates in a waveguide not for tens or hundreds of microns but many cm.  Can EPOCH run in such moving frame permanently adding mass in front of laser beam and deleting it at its back ?

Another way of solving such problems is Lorentz boosted frame method. How about it too?

**Comment by Status-Mirror** [comment]
Hey @DanRRRR,

This can be achieved using the EPOCH [window block](https://epochpic.github.io/documentation/input_deck/input_deck_window.html).

Cheers,
Stuart

**Comment by DanRRRR** [comment]
Well, i got the chance to continue my eternal rant again. About these freaking demo examples.  :).  My experience tells that without small demo examples 99% of things do not go without rivers of blood and lost days and weeks. I tried this method and here i constructed some artificial very small demo which runs seconds even on a single core. Demo ends immediately with error crash. I couldn't find any solution for it.

The other darn thing is that when i remove the moving frame block "begin:window" completely, all initially works up to the time 10 (100fs) and then code crashes with segmentation fault. 

There exist also one more curious problem in settings of density profile which literally pulled my last hair, it is marked " <-- if use this code crash instead". Use it instead of previous line above. Both these lines set some small density depression in the center by little bit different ways. I do not know why marked method is rejected even before the run   ;-(((((

All that confirms that leaving users without working demos is like dropping them in the middle on the night in deep forest full of traps, holes, furious bugs and animals

```

begin:control

  nx =  320
  ny =  320

# Final time of simulation

  t_end = 300 * femto

# Size of domain
  x_min = -16. * micron
  x_max =  16. * micron
  y_min = -16. * micron
  y_max =  16. * micron

end:control

#-------------------------------------
begin:boundaries
  bc_x_min = simple_laser
  bc_x_max = simple_outflow
  bc_y_min = simple_outflow
  bc_y_max = simple_outflow
end:boundaries

begin:window
   move_window = T
   window_v_x = 3.0e8
   window_start_time = 70.e-15
   bc_x_min_after_move = simple_outflow
   bc_x_max_after_move = simple_outflow
end:window

#-----------------------------------------------------------
begin:constant
  laser_wavelength = 0.8 * micron
  omega = 2 * pi * c / laser_wavelength
  n_crit = critical(omega)
  den_gas = 0.00005 * n_crit
  Hcharge  = 1

  GradL = 10. *micron
  HoleDep = 1.

  r = sqrt(y^2 + z^2)
#  z = 0
  laser_fwhm   = 3 * micron
  laser_dur_fwhm = 5.e-15

end:constant

 begin:collisions
  use_collisions = T
  use_nanbu = T
  coulomb_log = auto
# collisional_ionisation = T
# collide = electron electron on     #this key to switch off specific collisions because USE COLLISIONS=T means they are all ON by default (besides coll.ionization which needs additional switch to turn ON)
  coll_n_step = 10
  collide = all
end:collisions

#------------------------------------------------
begin:species
  name = Hydrogen
  charge = Hcharge
  mass   = 1836.2

  number_density = if( (x gt -100.) and (x lt 100.), den_gas/(1+HoleDep*exp(-(y/GradL)^2) ),  0.)
# number_density = if( (x gt -100.) and (x lt 100.), den_gas*(1.+(y/gradL)^2), 0.)                          <-- if use this code crash instead 

  npart_per_cell = 1
end:species


#---------------------------------------------------
begin:species
  name = electron
  charge = -1.0
  mass = 1.0
  number_density = number_density(Hydrogen)*Hcharge

  npart_per_cell = 1

end:species

#--------------------------------------------------

begin:laser
  boundary = x_min
  intensity_w_cm2 =  1.e19
  profile = gauss(y, 0., laser_fwhm)
  t_profile = gauss(time, 1.*laser_dur_fwhm, laser_dur_fwhm * (2.0 * sqrt(loge(2.0))))
  pol_angle = 0. # 0.5*pi      # S-polarization
  lambda = laser_wavelength
end:laser


#====================== O U T P U T =========================

begin:output_global
  force_final_to_be_restartable = T
end:output_global

#begin:output
#  name = restart
#  file_prefix = restart
#  restartable = T
#  dt_snapshot = 1.0e-12
#end:output


begin:output
  name = normal
  file_prefix = q
  dt_snapshot = 10 * femto
  grid = always
  ey = always

end:output```

**Reply by Status-Mirror** [reply]
Hey Dan,

Let's see if we can get you out of this forest.

Firstly, your demo input deck here is fine. I've removed the 3 \`\`\` after your `end:output`, and the code runs without seg-fault using the version currently present on GitHub. The code also runs to completion when I remove the window block. Your final hair-pulling issue is caused because EPOCH variables are case-sensitive: your working line has `GradL` and the broken line has the typo `gradL`.

On the greater subject of demos, I would like to start including more of these when I get free-time between development tasks. The prospect of creating a demo for every EPOCH feature is a little daunting (and not what I'm funded for), so I've mainly been targeting areas where people have expressed specific interest. From discussions on the issues page, I think the website would benefit from a moving window example, along with a discussion on the various ways to load density profiles, how to set-up an ionisation problem, and a little more information on particle injectors. Are there any other specific examples you would like to see?

Cheers,
Stuart

**Comment by DanRRRR** [comment]
Thanks, Stuart, for useful second opinion and found bugs with GradL/gradL. The Fortran rules which do not care about upper/lower case periodically give me such kick under. Another often made mistake is when i comment with an exclamation mark instead of # in input.deck (may be good idea to add it as comment ?  Exclamation mark also used as a comment in other languages).

Will try to catch the reason of my crashes with moving frame. 

I still have other general problems with EPOCH too, the restart for example works ok on one computer and one case and crash another computer or another case. What compiler/MPI and their versions do you use ?

As to what i'd see to add as the EPOCH examples - my respond would be to make the special topic where developers and users post their cases with pictures. Unlike Bill gates' "640K will be good for everyone", i'd guess 640 examples there would probably indeed fit everyone :)
